### ISSUE TITLE
26.1.0 Downloads.createDownload crash

### ISSUE THREAD

**Issue by Crazycatz00** [MoonchildProductions/Pale-Moon#369]
Since the 26.1.0 update, the `Downloads.createDownload` function from Downloads.jsm has begun failing.
This is because `_backendDownload` in Downloads.jsm has an argument named `aProperties`, but tries to access it as `aOptions`; additionally, it assumes `aProperties`\\`aOptions` is not `null`, when in this case it is.
It seems to be a simple fix, renaming `aProperties` to `aOptions` and setting it to an empty object if it's `null`.

Additionally, while this seems to be a fix, I'll mention it anyway. When passing a `nsIFile` to `createDownload` as the target, you previously had to pass `target: { file: nsiFile }`. Now it is simply `target: nsiFile`. The latter seems to be correct according to [Mozilla](https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Downloads.jsm#createDownload%28%29), but it is a breaking change.

**Comment by wolfbeast** [comment]
@chaoskagami Maybe you can have a look at this? This is your code.
I wanted to at least avoid breaking changes, but it seems one snuck in anyway?

**Comment by chaoskagami** [comment]
Ah, jeez. Thought I had ironed every possible issue out.

This shouldn't be too involved to fix - sorry about that. It's night here, so I'll fix it and PR in the morning. Hope that's okay. ;)

On February 18, 2016 5:21:17 AM EST, Moonchild <notifications@github.com> wrote:
>@chaoskagami Maybe you can have a look at this? This is your code.
>I wanted to at least avoid breaking changes, but it seems one snuck in
>anyway?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/MoonchildProductions/Pale-Moon/issues/369#issuecomment-185641593

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by chaoskagami** [comment]
Also, I was aiming for complete compatibility here, so I'm not entirely sure how I missed the nsiFile: {file: filename} bit. I'll fix that too.

On February 18, 2016 5:21:17 AM EST, Moonchild <notifications@github.com> wrote:
>@chaoskagami Maybe you can have a look at this? This is your code.
>I wanted to at least avoid breaking changes, but it seems one snuck in
>anyway?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/MoonchildProductions/Pale-Moon/issues/369#issuecomment-185641593

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by wolfbeast** [comment]
> I was aiming for complete compatibility here, so I'm not entirely sure how I missed the nsiFile: {file: filename} bit.

I think, if I understood correctly here, our old behavior needed { file: filename } but the new behavior doesn't? So that would mean our previous behavior changed (which I wanted to avoid). Please correct me if I misunderstood.

**Comment by Crazycatz00** [comment]
> I think, if I understood correctly here, our old behavior needed { file: filename } but the new behavior doesn't?

Correct. The exact code I'm using is
```
Downloads.createDownload({
	source: { uri: uri, referrer: uri, isPrivate: isPrivate },
	target: file,	// PM 26.1.0
	saver: { type: 'copy' }
})
```
where it used to be 
```
Downloads.createDownload({
	source: { uri: uri, referrer: uri, isPrivate: isPrivate },
	target: { file: file },	// PM 26.0.3 and below
	saver: { type: 'copy' }
})
```
I'm using a `nsILocalFile` as file, but same syntax.

From what I can tell, the old Pale Moon behavior is how FireFox 25 acted, while this new behavior is from FF 26. If the goal was newer API compatibility, some sort of argument sniffing may work to support both.

**Comment by chaoskagami** [comment]
As I said, that was a mistake. Both should be accepted, and I didn't intentionally break the api.

I'm getting to work now.

On February 19, 2016 2:47:26 PM EST, Crazycatz00 <notifications@github.com> wrote:
>> I think, if I understood correctly here, our old behavior needed {
>file: filename } but the new behavior doesn't?
>
>Correct. The exact code I'm using is
>```
>Downloads.createDownload({
>	source: { uri: uri, referrer: uri, isPrivate: isPrivate },
>	target: file,	// PM 26.1.0
>	saver: { type: 'copy' }
>})
>```
>where it used to be 
>```
>Downloads.createDownload({
>	source: { uri: uri, referrer: uri, isPrivate: isPrivate },
>	target: { file: file },	// PM 26.0.3 and below
>	saver: { type: 'copy' }
>})
>```
>I'm using a `nsILocalFile` as file, but same syntax.
>
>From what I can tell, the old Pale Moon behavior is how FireFox 25
>acted, while this new behavior is from FF 26. If the goal was newer API
>compatibility, some sort of argument sniffing may work to support both.
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/MoonchildProductions/Pale-Moon/issues/369#issuecomment-186374499

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by chaoskagami** [comment]
Should have fixed the crash and API break. Can you give it a check over?

**Comment by Crazycatz00** [comment]
Can confirm [d15e85ef4c16a2da4117954a8829b1e7da1578ae] has fixed the issue with `createDownload`, using both the FF25 and FF26 styles.

One final thing: When looking at the commit, I noticed `simpleDownload` uses `this._createDownload` (which doesn't exist) instead of `this._backendDownload`; this is causing an error as well.
Apart from that, the rest of Downloads.jsm seems to be fine, from the little I know of FF backend.

**Comment by chaoskagami** [comment]
I identified some other circumstances which cause things to fail, which is why I've rebased about seven times now. Can you force sync and check fd7a7f84 in my branch?

On February 19, 2016 6:27:26 PM EST, Crazycatz00 <notifications@github.com> wrote:
>Can confirm [d15e85ef4c16a2da4117954a8829b1e7da1578ae] has fixed the
>issue with `createDownload`, using both the FF25 and FF26 styles.
>
>One final thing: When looking at the commit, I noticed `simpleDownload`
>uses `this._createDownload` (which doesn't exist) instead of
>`this._backendDownload`; this is causing an error as well.
>Apart from that, the rest of Downloads.jsm seems to be fine, from the
>little I know of FF backend.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/MoonchildProductions/Pale-Moon/issues/369#issuecomment-186449462

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by chaoskagami** [comment]
Scratch that, try 763166d.

On February 19, 2016 6:27:26 PM EST, Crazycatz00 <notifications@github.com> wrote:
>Can confirm [d15e85ef4c16a2da4117954a8829b1e7da1578ae] has fixed the
>issue with `createDownload`, using both the FF25 and FF26 styles.
>
>One final thing: When looking at the commit, I noticed `simpleDownload`
>uses `this._createDownload` (which doesn't exist) instead of
>`this._backendDownload`; this is causing an error as well.
>Apart from that, the rest of Downloads.jsm seems to be fine, from the
>little I know of FF backend.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/MoonchildProductions/Pale-Moon/issues/369#issuecomment-186449462

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by Crazycatz00** [comment]
The latest commit seems to be working for me, for both FF25 and FF26 styles and both functions.

**Comment by chaoskagami** [comment]
Okay, good.

Hopefully 927a1f3 will be the end of the fixes, unless anyone else identifies anything severely wrong.

**Comment by wolfbeast** [comment]
What would be a good set of extensions to test this? e.g. one that crashed the browser, one that didn't work, etc.?

**Comment by chaoskagami** [comment]
Not all addons that download things FF25 style use Downloads.jsm. Many seem to directly use DownloadsCore or other available methods. For example; nicofox (https://github.com/littlebtc/nicofox) while requiring a GUID fix to display its' UI, directly issues SQLlite commands to the downloads db, it seems.

As for FF 26+ style, the following addon is an example of what DOES work now due to the changes:
https://addons.mozilla.org/en-US/firefox/addon/download-flash-and-video/

Some things are just plain broken though, due to reliance on other bits that I'm still trying to think of a way to make work. For example, Download StatusBar. S3 Download Manager's download functionality is broken, but there's deeper seated issues with it.

**Comment by tumpio** [comment]
I listed a couple of add-ons that expect the new API style from 26 version or higher in #326:
https://addons.mozilla.org/en-US/firefox/addon/download-flash-and-video/
https://addons.mozilla.org/en-US/firefox/addon/flash-video-downloader/
possibly IMD-addon forum.palemoon.org/viewtopic.php?f=16&t=10835
and the https://addons.mozilla.org/en-US/firefox/addon/imacros-for-firefox/ itself.

After the Pale Moon 26.1.0 release, this addon was reported to broke due to Download.jsm changes:
https://addons.mozilla.org/en-US/firefox/addon/complete-youtube-saver/

**Comment by wolfbeast** [comment]
If this can fix iMacros, that would be great! People have been bucking hard about this incompatibility since v26.

**Comment by Ordinaryperson2** [comment]
This patch seems to have partially fixed Complete Youtube Saver (no longer throws an error regarding aOptions) but the download still doesn't start, is there something else involved here? It works fine on 26.0.3 and earlier so it shouldn't be related to the version 26 check that other extensions tripped over.

**Comment by chaoskagami** [comment]
Link me the extension, please. I'm going to need to rip it apart for debugging.

As far as reading through my code, I can't see any other logical issues, so if it still isn't working correctly, then I need to do some serious investigation.

If I still can't figure out what's going wrong, my branch may need to be reverted temporarily (again.)

On February 25, 2016 12:25:06 AM EST, Poke <notifications@github.com> wrote:
>This patch seems to have partially fixed Complete Youtube Saver (no
>longer throws an error regarding aOptions) but the download still
>doesn't start, is there something else involved here? It works fine on
>26.0.3 and earlier so it shouldn't be related to the version 26 check
>that other extensions tripped over.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/MoonchildProductions/Pale-Moon/issues/369#issuecomment-188620926

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by Ordinaryperson2** [comment]
Here:
https://addons.mozilla.org/en-US/firefox/addon/complete-youtube-saver/

To test, go to a youtube video and click CYS's icon. Choose the quality you want, click save, and the download(s) are meant to start.

**Comment by chaoskagami** [comment]
Okay, looks like that addon isn't working for an entirely different reason than the old API being broken.

They appear to check whether the 'getList' function is supported, and if so, use the FF 26+ API, otherwise they fallback to the old API. At least they aren't doing pointless version checks.

This is triggered because the addon tries to use the new API thinking it's supported - it isn't completely. It starts off fine, then it attempts to add a new-style observer which isn't yet shimmed.

tl;dr - Not fixable unless the new API is removed, also not my problem. They're not checking support correctly.

**Comment by Ordinaryperson2** [comment]
Oh wow, thanks for looking into this, and sharing the specifics. Sounds like it should be a simple fix on the extension's end.

**Comment by chaoskagami** [comment]
They're not updating it anymore, but if you want to fix the offender, cysCommons.js line 420. Replace with `if(false){` to force the old API. Or add a check to the end like `&& xulAppInfo.name != 'Pale Moon'`.

**Comment by wolfbeast** [comment]
> Or add a check to the end like `&& xulAppInfo.name != 'Pale Moon'`

That would be a proper fix.

**Comment by chaoskagami** [comment]
If the addon could be redistributed, I would say so. ;)

Unfortunately the license is 'all rights reserved' on that, so it doesn't really matter.

**Comment by wolfbeast** [comment]
Well then, our hands are tied.
